sw/device/lib/otbn.h

```diff
@@ -0,0 +1,248 @@
+ */
+#define OTBN_USE_APP(app_name)                            \
+  extern const char _otbn_app_##app_name##__imem_start[]; \
+  extern const char _otbn_app_##app_name##__imem_end[];   \
+  extern const char _otbn_app_##app_name##__dmem_start[]; \
+  extern const char _otbn_app_##app_name##__dmem_end[];
+
+/**
+ * Initialize the OTBN application information structure
+ *
+ * After making all required symbols from the application image available
+ * through `OTBN_USE_APP()`, use this macro to initialize an `otbn_app_t`
+ * struct with those symbols.
+ *
+ * @param app_name name of the application to load
+ * @see OTBN_USE_APP()
+ */
+#define OTBN_APP_T_INIT(app_name)                     \
+  {                                                   \
+    .imem_start = _otbn_app_##app_name##__imem_start, \
+    .imem_end = _otbn_app_##app_name##__imem_end,     \
+    .dmem_start = _otbn_app_##app_name##__dmem_start, \
+    .dmem_end = _otbn_app_##app_name##__dmem_end,     \
+  }
+
+/**
+ * Make a symbol in the OTBN application image available.
+ *
+ * Symbols are typically function or data pointers, i.e. labels in assembly
+ * code.
+ *
+ * @param app_name    name of the application the function is contained in
+ * @param symbol_name name of the symbol (function, label)
+ */
+#define OTBN_USE_PTR(app_name, symbol_name) \
+  extern const char _otbn_app_##app_name##_##symbol_name[];
+
+/**
+ * Initialize an `otbn_ptr_t` structure.
+ */
+#define OTBN_PTR_T_INIT(app_name, symbol_name) \
+  { .ptr_in_cpu_addr_space = _otbn_app_##app_name##_##symbol_name, }
+
+/**
+ * Initialize the OTBN context structure.
+ *
+ * @param ctx the context object
+ * @param dif_config the OTBN DIF configuration (passed on to the DIF)
+ * @return #kOtbnOk if the operation was successful,
+ *         #kOtbnBadArg if a precondition check failed,
+ *         #kOtbnError otherwise
+ */
+otbn_result_t otbn_init(otbn_t *ctx, const dif_otbn_config_t *dif_config);
+
+/**
+ * (Re-)load the RSA application into OTBN.
+ *
+ * Load the application image with both instruction and data segments into OTBN.
+ *
+ * @param ctx the context object
+ * @param app the application to load into OTBN
+ * @return #kOtbnOk if the operation was successful,
+ *         #kOtbnBadArg if a precondition check failed,
+ *         #kOtbnError otherwise.
+ */
+otbn_result_t otbn_load_app(otbn_t *ctx, const otbn_app_t *app);
+
+/**
+ * Call a function on OTBN.
+ *
+ * Set the entry point (start address) of OTBN to the desired function, and
+ * starts the OTBN operation.
+ *
+ * Use `otbn_busy_wait_for_done()` to wait for the function call to complete.
+ *
+ * @param ctx the context object
+ * @param func the function to be called
+ * @return #kOtbnOk if the operation was successful,
+ *         #kOtbnFuncNotInApp if `func` isn't part of the loaded application,
+ *         #kOtbnError otherwise.
+ */
+otbn_result_t otbn_call_function(otbn_t *ctx, const otbn_ptr_t *func);
```

Comment:
otbn_ptr seems like a value type. Why is it taken by pointer?

Reply:
Switched to pass-by-value.